File: backend/src/main/java/com/colegio/bastidas/service/DocenteService.java
package com.colegio.bastidas.service;

import com.colegio.bastidas.dto.docente.DocenteCreadoResponseDTO;
import com.colegio.bastidas.dto.docente.DocenteDocumentoResponseDTO;
import com.colegio.bastidas.dto.docente.DocenteRequestDTO;
import com.colegio.bastidas.dto.docente.DocenteResponseDTO;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Optional;

/**
 * Contrato de negocio para la gestión de Docentes.
 */
public interface DocenteService {

    /** Lista todos los docentes con su estado curricular calculado. */
    List<DocenteResponseDTO> listarTodos();

    /** Busca un docente por su ID. */
    DocenteResponseDTO buscarPorId(Long id);

    /** Busca un docente por DNI. */
    DocenteResponseDTO buscarPorDni(String dni);

    /**
     * Crea un nuevo docente y aprovisiona automáticamente su cuenta de Usuario
     * (rol DOCENTE, contraseña inicial predecible que debe cambiar en su primer login).
     */
    DocenteCreadoResponseDTO crear(DocenteRequestDTO dto);

    /** Actualiza datos de un docente existente. */
    DocenteResponseDTO actualizar(Long id, DocenteRequestDTO dto);

    /** Da de baja (o reactiva) a un docente desactivando su cuenta de Usuario (ya no puede iniciar sesión). */
    DocenteResponseDTO actualizarEstadoActivo(Long id, boolean activo);

    /**
     * Semáforo Curricular: lista los 68 docentes con su estado
     * APROBADO | PENDIENTE | RETRASADO calculado por el sistema.
     */
    List<DocenteResponseDTO> obtenerSemaforoCurricular();

    /** Contadores para KPIs del semáforo: {aprobados, pendientes, retrasados}. */
    long contarPorEstadoCurricular(String estado);

    /** Exporta la lista de docentes con sus cursos asignados en formato Excel. */
    byte[] exportarListaConCursos();

    /** Exporta la lista de docentes con sus cursos asignados en formato PDF. */
    byte[] exportarListaConCursosPdf();

    /**
     * Sube (o reemplaza) el documento que certifica que el docente tiene o está
     * cursando docencia. No es obligatorio. Solo admite PDF.
     */
    String cargarDocumentoCertificacion(Long docenteId, MultipartFile archivo);

    /** Documento de certificación docente actual, si fue subido. */
    Optional<DocenteDocumentoResponseDTO> obtenerDocumentoCertificacion(Long docenteId);

    /** Elimina el documento de certificación docente, si existe. */
    void eliminarDocumentoCertificacion(Long docenteId);
}

File: frontend/src/app/core/services/docente.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface DocenteResponse {
  id: number;
  codigoDocente: string;
  dni: string;
  apellidoPaterno: string;
  apellidoMaterno: string;
  nombres: string;
  nombreCompleto: string;
  especialidad: string;
  condicion: 'NOMBRADO' | 'CONTRATADO';
  emailInstitucional: string;
  celular: string;
  estadoCurricular: 'APROBADO' | 'PENDIENTE' | 'RETRASADO';
  cantidadEvidencias: number;
  cantidadCursosAsignados: number;
  activo: boolean;
  tieneCertificacionDocencia: boolean;
}

export interface SemaforoConteo {
  aprobados: number;
  pendientes: number;
  retrasados: number;
}

export interface DocenteRequestDTO {
  dni: string;
  apellidoPaterno: string;
  apellidoMaterno: string;
  nombres: string;
  especialidad?: string;
  condicion?: 'NOMBRADO' | 'CONTRATADO';
  emailInstitucional?: string;
  celular?: string;
}

export interface DocenteCreadoResponse {
  docente: DocenteResponse;
  usernameGenerado: string;
  passwordInicial: string;
}

export interface CursoAsignado {
  id: number;
  areaCurricular: string;
  anioAcademico: number;
  aula: {
    id: number;
    grado: string;
    seccion: string;
    nivel: string;
  };
  docente?: {
    nombres: string;
    apellidoPaterno: string;
    apellidoMaterno: string;
    nombreCompleto: string;
  };
  diaSemana?: 'LUNES' | 'MARTES' | 'MIERCOLES' | 'JUEVES' | 'VIERNES';
  horaInicio?: string; // "HH:mm:ss"
  horaFin?: string;
}

/**
 * Servicio Angular para consumir la API de Docentes.
 * Base: /api/v1/docentes
 */
@Injectable({
  providedIn: 'root'
})
export class DocenteService {

  private readonly BASE = `${environment.apiUrl}/docentes`;

  constructor(private http: HttpClient) {}

  /** Lista todos los docentes. */
  listarTodos(): Observable<DocenteResponse[]> {
    return this.http.get<DocenteResponse[]>(this.BASE);
  }

  /** Obtiene la lista completa con estado del Semáforo Curricular. */
  obtenerSemaforo(): Observable<DocenteResponse[]> {
    return this.http.get<DocenteResponse[]>(`${this.BASE}/semaforo`);
  }

  /** KPIs de conteo del Semáforo para el Dashboard. */
  obtenerConteosemaforo(): Observable<SemaforoConteo> {
    return this.http.get<SemaforoConteo>(`${this.BASE}/semaforo/conteo`);
  }

  /** Busca un docente por ID. */
  buscarPorId(id: number): Observable<DocenteResponse> {
    return this.http.get<DocenteResponse>(`${this.BASE}/${id}`);
  }

  /** Busca un docente por su DNI. */
  buscarPorDni(dni: string): Observable<DocenteResponse> {
    return this.http.get<DocenteResponse>(`${this.BASE}/dni/${dni}`);
  }

  /** Obtiene los cursos asignados a un docente */
  obtenerCursos(id: number): Observable<CursoAsignado[]> {
    return this.http.get<CursoAsignado[]>(`${this.BASE}/${id}/cursos`);
  }

  /** Registra un nuevo docente; el backend crea su cuenta de usuario automáticamente. */
  crear(dto: DocenteRequestDTO): Observable<DocenteCreadoResponse> {
    return this.http.post<DocenteCreadoResponse>(this.BASE, dto);
  }

  /** Actualiza los datos de un docente existente (no cambia DNI). */
  actualizar(id: number, dto: DocenteRequestDTO): Observable<DocenteResponse> {
    return this.http.put<DocenteResponse>(`${this.BASE}/${id}`, dto);
  }

  /** Da de baja (o reactiva) al docente: desactiva su cuenta, ya no puede iniciar sesión. */
  actualizarEstado(id: number, activo: boolean): Observable<DocenteResponse> {
    return this.http.patch<DocenteResponse>(`${this.BASE}/${id}/estado`, {}, { params: { activo: String(activo) } });
  }

  /** Descarga la lista de docentes con sus cursos asignados en Excel. */
  exportar(): Observable<Blob> {
    return this.http.get(`${this.BASE}/exportar`, { responseType: 'blob' });
  }

  /** Descarga la lista de docentes con sus cursos asignados en PDF. */
  exportarPdf(): Observable<Blob> {
    return this.http.get(`${this.BASE}/exportar/pdf`, { responseType: 'blob' });
  }
}
